plutus-tx/src/PlutusTx/Bool.hs

```diff
@@ -37,3 +37,28 @@ infixr 2 ||
 --
 not :: Bool -> Bool
 not a = if a then False else True
+
+{- Note [Lazy patterns on function parameters]
+Lazy patterns (~) on function parameters don't do anything if the function isn't inlined.
+This is because function applications in Plutus Tx are strict, so when passing an argument
+to a function, the argument will be evaluated immediately, regardless of whether the
+corresponding function parameter is lazy or strict. Specifically, both `f !x = body` and
+`f ~x = body` in Plutus Tx are compiled into `let f = \x -> body` in PIR.
+
+It doesn't matter whether or not the function is inlined by the PIR inliner, since the PIR
+inliner does not change the semantics of the program.
+
+However, it *does* make a difference if the function is unconditionally inlined by GHC.
+Consider the Plutus Tx code `let f !x = t in f arg`. Here `arg` will be immediately when
+being passed to `f` (as it should). But now consider `let f ~x = t in f arg`. Here GHC
+may unconditionally inline `f`, leading to `t [x := arg]` (recognize that the GHC inliner
+also performs beta reduction, in addition to replacing `f` with its definition), where `arg`
+may not be evaluated. GHC does so because it is a Haskell compiler and does not know that
+Plutus Tx is a strict language.
+
+Therefore, lazy patterns should generally be avoided, since its semantics depend on
+what the GHC inliner does. That said, we can use it in some cases to our advantage,
+such as in `(&&)` and `(||)`. These logical operators would be much less useful if
+they can't short-circuit. By using a lazy pattern on the second parameter, we can achieve
```

Comment:
I disagree with this. The INLINE pragma does not force GHC to inline. We run the GHC simplifier with `sm_inline` turned off. It does force the PIR inliner to inline, but the PIR inliner only replaces the variable with the definition, which isn't sufficient to make it non-strict. We should make it so that the INLINE pragma forces the PIR simplifier to beta-reduce. We had a length discussion in the ADR PR and IIRC this was the conclusion.

Reply:
You are right yes, we need to both force inlining and beta reduction. I keep forgetting that!

> We had a length discussion in the ADR PR and IIRC this was the conclusion.

I do not think we agreed to that :p https://github.com/input-output-hk/plutus/pull/5270#discussion_r1185970670

I'm overall then confused by this line: since we don't have such behaviour at the moment, we _can't_ achieve what's written here? So the note is talking about some hypothetical future behaviour. I think we should be clear that it's not something we can do now!

Reply:
Agreed. I rephrase it in an attempt to make it more clear:

> Using a lazy pattern on the second parameter helps achieve
the desirable effect, namely, the second parameter is only evaluated if needed, although
this is currently not guaranteed, since `(&&)` and `(||)` are not necessarily
unconditionally inlined by GHC. To guarantee that the second parameter is only evaluated
when needed, we can potentially use a particular source annotation to tell the PIR
inliner to inline and beta-reduce a function, which would achieve the same effect as if
it is unconditionally inlined by GHC. This is yet to be implemented.